Add checkBlock() to Mining interface

Sjors opened the pull request:
This PR adds the IPC equivalent of the `getblocktemplate` RPC in `proposal` mode.

In order to do so it has `TestBlockValidity` return error reasons as a string instead of `BlockValidationState`. This avoids complexity in IPC code for handling the latter struct.

The new Mining interface method is used in `miner_tests`.

It's not used by the `getblocktemplate` and `generateblock` RPC calls, see https://github.com/bitcoin/bitcoin/pull/31981#discussion_r2096473337

The `inconclusive-not-best-prevblk` check is moved from RPC
code to `TestBlockValidity`.

Test coverage is increased by `mining_template_verification.py`.

Superseedes #31564

## Background

### Verifying block templates (no PoW)

Stratum v2 allows miners to generate their own block template. Pools may wish (or need) to verify these templates. This typically involves comparing mempools, asking miners to providing missing transactions and then reconstructing the proposed block.[^0] This is not sufficient to ensure a proposed block is actually valid. In some schemes miners could take advantage of incomplete validation[^1].

The Stratum Reference Implementation (SRI), currently the only Stratum v2 implementation, collects all missing mempool transactions, but does not yet fully verify the block.[^2]. It could use the `getblocktemplate` RPC in `proposal` mode, but using IPC is more performant, as it avoids serialising up to 4 MB of transaction data as JSON.

(although SRI could use this PR, the Template Provider role doesn't need it, so this is _not_ part of #31098)

[^0]: https://github.com/stratum-mining/sv2-spec/blob/main/06-Job-Declaration-Protocol.md  
[^1]: https://delvingbitcoin.org/t/pplns-with-job-declaration/1099/45?u=sjors
[^2]: https://github.com/stratum-mining/stratum/blob/v1.1.0/roles/jd-server/src/lib/job_declarator/message_handler.rs#L196

Sjors committed:
ipc: drop BlockValidationState special handling

The Mining interface avoids using BlockValidationState.

Sjors commented:
I also salvaged the functional tests from 15d841e41bb61f435dd3e60e99fa4db47d34a3a6.

Sjors commented:
Rebased after #31283.

Sjors commented:
Rebased after #31974 (trivial conflict with 3b33fd9424a2b4edbdc1745339e3541ab4b90af5).

l0rinc commented on src/node/types.h:
```diff
@@ -85,6 +86,17 @@ struct BlockWaitOptions {
 
+struct BlockCheckOptions {
+    /**
+     * Set false to omit the merkle root heck
```
Can we turn this off?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31981#discussion_r2033995107

I actually requested the review since I was curious what it would do. It seems a bit afraid to be opinionated though and might be more useful if we added guidelines https://docs.github.com/en/enterprise-cloud@latest/copilot/using-github-copilot/code-review/configuring-coding-guidelines

Sjors replied:
Fixed typo since I had to retouch anyway.

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -740,13 +740,9 @@ static RPCHelpMan getblocktemplate()
-            TestBlockValidity(state, chainman.GetParams(), chainman.ActiveChainstate(), block, chainman.m_blockman.LookupBlockIndex(block.hashPrevBlock), /*fCheckPOW=*/false, /*fCheckMerkleRoot=*/true);
-            return BIP22ValidationResult(state);
+            std::string reason;
+            bool res{chainman.TestBlockValidity(block, reason, /*check_pow=*/false, /*check_merkle_root=*/true)};
```
In commit "validation: refactor TestBlockValidity" (1d029c23a143f1b9110377967a7c4a20ee75058d)

it seems like there could be change in behavior here since BIP22ValidationResult would previously throw an exception on error and now it will just return an error string.

Sjors replied:
Good point. The original code would return a string in some cases and throw an error in others. Although it's possible nobody uses the `getblocktemplate` RPC in `proposal` mode at all, there could be (unmaintained) proprietary software out there using it. So the safest option is to maintain the behavior.

However it seems there's no change in behavior: `BIP22ValidationResult` only throws for `state.IsError()`, not for `state.IsInvalid()`. I looked through `CheckBlock` and the functions it calls and none of those can set `state.IsError()`.

I expanded the commit message to explain this.

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -386,9 +386,9 @@ static RPCHelpMan generateblock()
-            throw JSONRPCError(RPC_VERIFY_ERROR, strprintf("TestBlockValidity failed: %s", state.ToString()));
+        std::string reason;
+        if (!chainman.TestBlockValidity(block, reason, /*check_pow=*/false, /*check_merkle_root=*/false)) {
+            throw JSONRPCError(RPC_VERIFY_ERROR, strprintf("TestBlockValidity failed: %s", reason));
```
In commit "validation: refactor TestBlockValidity" (1d029c23a143f1b9110377967a7c4a20ee75058d)

There seems to be a change in behavior here if `ValidationState::m_debug_message` is set because now it would previous be included in `ToString` output but now will be dropped.

Sjors replied:
I replaced the four `state.GetRejectReason()` with `state.ToString()`, as it was in the original.

Sjors replied:
I changed `TestBlockValidity` to also take a `debug` argument.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4641,40 +4641,86 @@ MempoolAcceptResult ChainstateManager::ProcessTransaction(const CTransactionRef&
-                       CBlockIndex* pindexPrev,
-                       bool fCheckPOW,
-                       bool fCheckMerkleRoot)
+bool ChainstateManager::TestBlockValidity(const CBlock& block, std::string& reason, const bool check_pow, const bool check_merkle_root)
```
In commit "validation: refactor TestBlockValidity" (1d029c23a143f1b9110377967a7c4a20ee75058d)

Most of the changes here seem good, but there are a few things I think could be improved:

- It seems better to me if this would return a `BlockValidationState` object instead of a string and a bool, so the valid/invalid/error states could be clearly distinguished, debug information could be returned, and this would fit in better with other `validation.h` functions using `BlockValidationState`. Returning `BlockValidationState` would also let callers use the `BIP22ValidationResult` univalue conversion function so we could reduce inconsistencies in how RPCs return validation information.
- Also think it would be good if this function still took a `chainstate` parameter, since it shouldn't care about which chainstate is active and could work on any chainstate.
- I think it would be better if this remained a standalone function instead of becoming a ChainstateManager method since it shouldn't need access to private ChainstateManager state and we might want to move it out of `validation.cpp` at some point to someplace like `node/miner.cpp` since it is only used by mining code.
- Probably also better for it not to `LOCK(cs_main)` to avoid unnecessary recursive mutex locking.
- Using `Assert()` consistently here would seem better than using a mix of `Assert` and `Assume`. Or if `Assume()` should be used it would be better to check return value and return an error status, instead of the ignoring failures in release builds.

Made a commit with all these changes which looks like: bf15692ab9b5f8d68cb84e925c067caebe03efba ([branch](https://github.com/ryanofsky/bitcoin/commits/pr/review.31981.1-edit))

Sjors replied:
Making this return a `BlockValidationState` means we have to pass it over the interface, which means we can't drop the special handling, see cca5993b26e6223af31fe1ef5bf8a319cb87cf93.

Sjors replied:
None of the `Assume` checks seem problematic in production. They would just be inconsistent with (my understanding of) previous behavior. I dropped them, since none of the tests and fuzzers tripped over them.

I dropped the `!block.fChecked `Assert`. The remaining `Assert`s are where we would crash anyway.

Finally I changed the pre-existing `assert` at the end into an Assume along with a LogError.

ryanofsky reviewed (approved):
Code review ACK e6e170cf6c67a56b9c14cece66fdc4fab5f3ec6b. Change is nicely implemented and background in the PR description is very helpful. Am a little concerned about possible changes in RPC behavior though and left some comments below.

Sjors commented:
Note to self, figure out if the sigops check needs to be fixed here or in another PR: https://github.com/bitcoin/bitcoin/pull/31624/files#r2053597806

Sjors commented:
Rebased, fixed typo found by LLM, clarified non behavior change for `proposal` mode.

I added a `debug` argument to pass along more detailed failure reasons.

I'm a bit on the fence regarding this suggestion: https://github.com/bitcoin/bitcoin/pull/31981#discussion_r2035713597

I also still need to look at the sigops check.

Sjors committed:
validation: refactor TestBlockValidity

Comments are expanded.

Return BlockValidationState instead of passing a reference.
Lock Chainman mutex instead of cs_main.
Remove redundant chainparams and pindexPrev arguments.
Drop defaults for checking proof-of-work and merkle root.

The ContextualCheckBlockHeader check is moved to after CheckBlock,
which is more similar to normal validation where context-free checks
are done first.

Validation failure reasons are no longer printed through LogError(),
since it depends on the caller whether this implies an actual bug
in the node, or an externally sourced block that happens to be invalid.
When called from getblocktemplate, via BlockAssembler::CreateNewBlock(),
this method already throws an std::runtime_error if validation fails.

Additionally it moves the inconclusive-not-best-prevblk check from RPC
code to TestBlockValidity.

There is no behavior change when callling getblocktemplate with proposal.
Previously this would return a BIP22ValidationResult which can throw for
state.IsError(). But CheckBlock() and the functions it calls only use
state.IsValid().

The final assert is changed into Assume, with a LogError.

Co-authored-by: <Ryan Ofsky <ryan@ofsky.org>

Sjors commented:
Forgot to adjust the capnp file. Fixed more typos.

I moved `TestBlockValidity` out of `ChainstateManager` again, and pass in a `chainstate` parameter as suggested in https://github.com/bitcoin/bitcoin/pull/31981#discussion_r2035713597.

> Probably also better for it not to `LOCK(cs_main)` to avoid unnecessary recursive mutex locking.

Can you elaborate on this?

davidgumberg commented on src/validation.cpp:
```diff
@@ -4648,40 +4648,82 @@ MempoolAcceptResult ChainstateManager::ProcessTransaction(const CTransactionRef&
+    index_dummy.pprev = tip;
+    index_dummy.nHeight = tip->nHeight + 1;
+    index_dummy.phashBlock = &block_hash;
+    CCoinsViewCache tip_view(&chainstate.CoinsTip());
+    CCoinsView blockCoins;
+    CCoinsViewCache view(&blockCoins);
+    view.SetBackend(tip_view);
```
```suggestion
    CCoinsViewCache view_dummy(&chainstate.CoinsTip());
```

Sjors replied:
I'm confused about what you're suggesting here.

TheCharlatan replied:
In commit 981853ab25dc6f967f393ecba57965afb8a822f3

I would have suggested the same. It is not clear to me why you are introducing a new `blockCoins` (camel case?) view here. Maybe do some of these `TestBlockValidity` refactors and documentation in a separate commit. It is not quite clear to me why this should be changed.

Sjors replied:
This doesn't compile, but I assume you also meant to pass `view_dummy` into the `ConnectBlock` call below? I don't remember what the original was based on, so took the suggestion.

Sjors replied:
At least we have a test that checks the UTXO set isn't updated: https://github.com/bitcoin/bitcoin/blob/1def3bc3773b3013a92f54daa75acb82c30beaac/test/functional/mining_template_verification.py#L145-L156

But that doesn't cover all conceivable future ways that we might accidentally flush.

If we do want to add a belt-and-suspenders way to prevent such a flush, it seems better for a followup.

> there could be better ways of preventing it like adding a read-only option to `CCoinsViewCache` that disables writing to the base UTXO set

That seems like the best approach.

stringintech commented on src/validation.cpp:
```diff
@@ -4648,40 +4648,82 @@ MempoolAcceptResult ChainstateManager::ProcessTransaction(const CTransactionRef&
-        LogError("%s: Consensus::ContextualCheckBlockHeader: %s\n", __func__, state.ToString());
+    // Lock must be held throughout this function for two reasons:
+    // 1. We don't want the tip to change during several of the validation steps
+    // 2. To prevent a CheckBlock() race condition for fChecked, see ProcessNewBlock()
```
Since `TestBlockValidity` is typically called with a fresh block instance, I'm not seeing how the race condition with `fChecked` could occur. Could you elaborate on this please?

Sjors replied:
I think you're right that this can't happen in the current codebase. `TestBlockValidity` is only called by `checkBlock()`. When this is called via the IPC interface, it will have a fresh `CBlock` each time (unless we start supporting shared memory between processes). Simiarly when called by the `generateblock` RPC or `getblocktemplate` in `proposal` mode it will have a fresh instance.

The code in `miner_tests.cpp` does reuse the same `CBlock` across multiple calls, though not in parallel.

Still, it seems like a good precaution against (perhaps unlikely) future regressions.

stringintech replied:
Thank you for the clarification.

ismaelsadeeq reviewed (commented):
Concept ACK after https://bitcoincore.reviews/31981

Sjors commented:
Thanks. I still need to address the questions by @stringintech.

ryanofsky commented on test/functional/mining_template_verification.py:
```diff
@@ -0,0 +1,274 @@
+    def set_test_params(self):
+        self.num_nodes = 1
+
+    def run_test(self):
```
In commit "test: more template verification tests" (c1939c43c3addb17c4316d49580762a1e0ec4504)

Might be nice to split this test up into different methods since most of the checks seem independent. (Could pass in any objects they are sharing in common as arguments)

Sjors replied:
Done (though I didn't put too much effort into making them actually independent).

ryanofsky reviewed (approved):
Code review ACK c1939c43c3addb17c4316d49580762a1e0ec4504. Since last review changes were improving error handling adding many comments, and making TestBlockValidity a standalone method again accepting a chainstate.

I left another suggestion (and a diff) below to make TestBlockValidity return a BlockValidationState instead of a bool and strings, which I think would be better, but feel free to keep current approach if you prefer.

Might also be nice to link to the review club discussion in the description (https://bitcoincore.reviews/31981)

Sjors commented:
I took @ryanofsky's patch to keep BlockValidationState in the interface _implementation_ and validation code.

Also split the test into multiple functions.

Sjors commented:
Rebased to include the new comment added in #31624.

The tests I added don't cover the difference in sigops counting. Suggestions are welcome, it could look for "CheckBlock failed" vs "ConnectBlock failed".

ryanofsky reviewed (approved):
Code review ACK c759ec1256a5cdc02b6eef1f4c0935c693bb8087. Since last review just rebased, applied some suggested changes to use BlockValidationState, and split the new python test into functions.

Sjors commented on test/functional/mining_template_verification.py:
```diff
@@ -0,0 +1,311 @@
+
+        self.current_tip_test(node, block_2)
+        self.nbits_test(node, block_2)
+        self.pow_test(node, block_2)
```
It seems Python doesn't believe in `const`, so I ended up just documenting the test and its call site.

Sjors committed:
Add checkBlock to Mining interface

Use it in miner_tests.

The getblocktemplate and generateblock RPC calls don't use this,
because it would make the code more verbose.

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -387,8 +387,10 @@ static RPCHelpMan generateblock()
 
-        if (BlockValidationState state{TestBlockValidity(chainman.ActiveChainstate(), block, /*check_pow=*/false, /*check_merkle_root=*/false)}; !state.IsValid()) {
-            throw JSONRPCError(RPC_VERIFY_ERROR, strprintf("TestBlockValidity failed: %s", state.ToString()));
+        std::string reason;
```
I dropped the changes to `rpc/mining.cpp` and explained why they're untouched in the commit message. Indeed `checkBlock` is already covered by the unit tests, and the meat and potatoes is in `TestBlockValidity`, which both IPC and RPC call.

Sjors replied:
(note: I'll push this in a few hours, just in case you're up early...)

Sjors commented:
Trivial rebase after #32562, dropped changes in RPC code and added a comment in the test.

ryanofsky reviewed (approved):
Code review ACK 975991571de3f11399ca01753e41753e0008b5a0. Since last review just dropped some RPC changes which were not strictly necessary and added test comments

sedited reviewed (commented):
Concept ACK

ryanofsky reviewed (approved):
Code review ACK 1def3bc3773b3013a92f54daa75acb82c30beaac. Only change since previous review is implementing a previous suggestion and passing a single cache temporary coinsview instead of a double cache to ConnectBlock.

Sjors commented on test/functional/mining_template_verification.py:
```diff
@@ -0,0 +1,313 @@
+)
+
+
+class MiningTemplateVerificationTest(BitcoinTestFramework):
```
It could make sense to depulicate things by moving a few things out of `mining_basic.py` here.

Sjors committed:
test: move gbt proposal mode tests to new file

Additionally this commit gives each test its
own function.

The assert_submitblock helper is absorbed into
assert_template.

Review hint:
git show --color-moved=dimmed-zebra

davidgumberg commented on src/validation.cpp:
```diff
@@ -4598,42 +4598,79 @@ MempoolAcceptResult ChainstateManager::ProcessTransaction(const CTransactionRef&
+
+    // For signets CheckBlock() verifies the challenge iff fCheckPow is set.
+    if (!CheckBlock(block, state, chainstate.m_chainman.GetConsensus(), /*fCheckPow=*/check_pow, /*fCheckMerkleRoot=*/check_merkle_root)) {
+        if (state.IsValid()) state.Error("CheckBlock failed");
```
I'm a bit confused by the `state.IsValid()` checks, in what circumstance would the checks fail but `state.IsValid()`?

Sjors replied:
This shouldn't happen, but if it does, we don't want to accidentally approve the block. I added `NONFATAL_UNREACHABLE` to make this clear.

TheCharlatan replied:
I think prefer how it is to be honest. Unreachable does communicate more clearly that this condition should never be reached.

Sjors commented:
Rebased after #32421.

Sjors commented:
Turns out I got a little carried away by #32421; blocks do still need `rehash()`.

ryanofsky reviewed (approved):
Code review ACK 2984e61ac0b51ccec46bb0159872becccbc4936c, just adding NONFATAL_UNREACHABLE since the last review and making a lot of nice changes to the tests including moving tests from mining_basic.py -> test/functional/mining_template_verification.py and using `assert_template` to make the tests more readable

davidgumberg commented on test/functional/mining_template_verification.py:
```diff
@@ -0,0 +1,302 @@
+    })
+    assert_equal(rsp, expect)
+    # Only attempt to submit invalid templates
+    if submit is not False and expect is not None:
```
nit, with above:

```suggestion
    if submit and expect is not None:
```

davidgumberg reviewed (approved):
utACK 2984e61ac0b51ccec46b

I'm not super familiar with how the IPC/capnproto schemas work, but the changes to `TestBlockValidity()` and the `getblocktemplate` tests look good to me.

Sjors commented:
I removed the setting of `state.Error` before `NONFATAL_UNREACHABLE`: https://github.com/bitcoin/bitcoin/pull/31981#discussion_r2144964512

Also took @davidgumberg's suggestions.

Both should improve readability.

ryanofsky reviewed (approved):
Code review ACK 8aad3bd81efd6ab70e78c60beedbfc425d609bf0 with minor NONFATAL_UNREACHABLE and assert_template improvements since last review

Sjors committed:
test: more template verification tests

sedited reviewed (approved):
ACK a18e57232867d946bc35769632fed49e1bf1464f

ryanofsky reviewed (approved):
Code review ACK a18e57232867d946bc35769632fed49e1bf1464f just adding another NONFATAL_UNREACHABLE since last review

achow101 commented:
ACK a18e57232867d946bc35769632fed49e1bf1464f
